lib/Dialect/TensorExt/Transforms/RotateAndReduce.cpp

```diff
@@ -187,7 +302,77 @@ struct RotateAndReduce : impl::RotateAndReduceBase<RotateAndReduce> {
+    // dataflow analysis, or else the analysis will be a no-op. Cf.
+    // https://github.com/llvm/llvm-project/issues/58922
+    solver.load<dataflow::DeadCodeAnalysis>();
+    solver.load<dataflow::SparseConstantPropagation>();
+    solver.load<rotation_analysis::RotationAnalysis>();
+
+    if (failed(solver.initializeAndRun(getOperation()))) {
+      getOperation()->emitOpError() << "Failed to run dataflow analysis.\n";
+      signalPassFailure();
+      return;
+    }
+
+    LLVM_DEBUG({
+      getOperation()->walk([&](Operation *op) {
+        if (op->getNumResults() == 0) return;
+        auto *targetSlotLattice =
+            solver.lookupState<rotation_analysis::RotationLattice>(
+                op->getResult(0));
+        if (targetSlotLattice->getValue().isOverdetermined()) {
+          llvm::dbgs() << "Rotation lattice for " << *op
+                       << " is overdetermined\n";
+        } else if (targetSlotLattice->getValue().empty()) {
+          llvm::dbgs() << "Rotation lattice for " << *op << " is empty\n";
+        } else {
+          SmallVector<int64_t> sortedRotations(
+              targetSlotLattice->getValue().getAccessedIndices().begin(),
+              targetSlotLattice->getValue().getAccessedIndices().end());
+          llvm::sort(sortedRotations);
+          std::string stringified = llvm::join(
+              llvm::map_range(sortedRotations,
+                              [](int64_t i) { return std::to_string(i); }),
+              ",");
+          llvm::dbgs() << "Rotation lattice for " << *op << ": " << stringified
+                       << "\n";
+        }
+      });
+    });
+
     DenseSet<Operation *> visited;
+
+    getOperation()->walk<WalkOrder::PreOrder, ReverseIterator>(
+        [&](Operation *op) {
+          if (op->getNumResults() == 0) return;
+          auto *targetSlotLattice =
+              solver.lookupState<rotation_analysis::RotationLattice>(
+                  op->getResult(0));
+          if (targetSlotLattice->getValue().isUninitialized() ||
+              targetSlotLattice->getValue().isOverdetermined()) {
+            return;
+          }
+
+          auto tensor = targetSlotLattice->getValue().getTensor();
+          auto accessIndices =
+              targetSlotLattice->getValue().getAccessedIndices();
+          int64_t tensorSize =
+              tensor.getType().cast<RankedTensorType>().getShape()[0];
+          if (accessIndices.size() == tensorSize) {
+            llvm::TypeSwitch<Operation &>(*op)
+                .Case<arith::AddIOp>([&](auto arithOp) {
+                  tryReplaceRotations<arith::AddIOp>(arithOp, tensor, visited,
+                                                     solver);
+                })
+                .Case<arith::MulIOp>([&](auto arithOp) {
+                  tryReplaceRotations<arith::MulIOp>(arithOp, tensor, visited,
+                                                     solver);
```

Comment:
i thank because this is iterating over each op, and then analyzing the backwards slice originating from the op, that you will eventually catch a tensor that has both a full-index-add and a full-index-mul. Could you just quickly add a test to simple_sum with a tensor that does both to make sure it does replace both of them?

Reply:
Done